File: 2017/day1/day1.py
from typing import List, Optional


class Captcha:
    def getInput(self) -> List[int]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        with open(inputFile, "r") as file1:
            return [int(x.strip()) for x in file1.readline()]

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.numList = self.getInput()

    def getSum(self, isPart2: Optional[bool] = False) -> int:
        length = len(self.numList)
        mid = length // 2
        prev = self.numList[mid] if isPart2 else self.numList[-1]
        total = 0
        for i in range(length):
            n = self.numList[i]
            if prev == n:
                total += n
            mid = (mid + 1) % length
            prev = self.numList[mid] if isPart2 else n
        return total


if __name__ == "__main__":
    captcha = Captcha()
    print("Day 1 part 1:", captcha.getSum())
    print("Day 1 part 2:", captcha.getSum(True))

File: 2019/day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

type Rocket struct {
	UseTest bool
	Masses  []int
}

func (this *Rocket) getInput() {
	inputFile := "input.txt"
	if this.UseTest {
		inputFile = "input-test.txt"
	}
	file, _ := os.Open(inputFile)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		mass, _ := strconv.Atoi(line)
		this.Masses = append(this.Masses, mass)
	}
	defer file.Close()
}

func (this *Rocket) getFuel(isPart2 bool) int {
	total := 0
	for _, num := range this.Masses {
		fuel := int(math.Floor(float64(num)/3) - 2)
		if isPart2 {
			for fuel > 0 {
				total += fuel
				fuel = int(math.Floor(float64(fuel)/3) - 2)
			}
		} else {
			total += fuel
		}
	}
	return total
}

func main() {
	rocket := &Rocket{
		UseTest: false,
		Masses:  []int{},
	}
	rocket.getInput()
	fmt.Println("Day 1 part 1:", rocket.getFuel(false))
	fmt.Println("Day 1 part 2:", rocket.getFuel(true))
}
